### PR TITLE
policy: bump TX_MAX_STANDARD_VERSION to 3

### PR DESCRIPTION
**Pull request by glozow**
Make `nVersion=3` (which is currently nonstandard on mainnet) standard.

Note that we will treat these transactions as Topologically Restricted Until Confirmation (TRUC). Spec is in BIP 431 and implementation is in #28948, #29306, and #29873

See #27463 for overall project tracking, and #29319 for information about relevance to cluster mempool.

### CODE REVIEW

**Comment by ariard** [comment]
What is your thinking on #29454 ?

I would still favor to make the 1000 vb limit dynamic and applying on the whole package, not only the child.
Doing with v3 itself would present a) a policy signaling mechanism saving and b) avoid some policy bargains like we had with `mempoolfullrbf` (between LN-vs-0-confs).

If you prefer to have the discussion on the v3 BIP PR let me know.

**Comment by ariard** [comment]
> What is your thinking on https://github.com/bitcoin/bitcoin/issues/29454 ?

Same as said that for sibling evictions - I’ll go to hack a tx-relay jamming / pinning toolkit, that way we’ll have more information on the robustness of those current proposed anti-pinning mitigations. I still maintain there are very incomplete for time-sensitive transaction flows, especially in face of “loophole” and NTA pinnings. I’m incredibly worried for the security of the Lightning Network if we don’t deploy additional anti-pinning mitigations on top of those ones over the future years - We have not seen yet script kiddies attacking the network. 

Under today inertia, we’ll probably converge towards a centralized network of ~50 LSPs relying on traditional social trust rather than relying on trust-minimized confirmation of pre-signed transactions. This is very problematic if they wish to accept channel traffic from low-resources spokes LN nodes, especially in geographical areas where social trust mechanism cannot be afforded by end-users.

On #29454, I still prefer to keep experimenting on “transaction-issuer selected policy limits” which might encompass more stringent tx-relay jamming on a minor fork of Bitcoin Core. Coming up with a sound technical proposal in matters of tx-relay policy likely demands few rounds of iterations and realistically we already have a severe shortage of qualified review bandwidth in those areas.

The initial reason closing of #29448 was my own realization that the introduction of differentiated tx-relay policy among the tx-relay network of nodes will probably generate with time tiers of transaction economic traffic priority (I really cannot see how it worsens memory / CPU DoS robustness per se ?). I don’t think this is a phenomena we can do a lot as a full-node implementation, as we’re seeing with the appearance of transaction accelerators by mining pools. There is transaction issuer demand for such services apparently. Yet we should be careful of how doing works in parallel to maintain bip152 block propagation and as such long-term decentralization of the base-layer network.

I’m not opposing the deployment of `TX_MAX_STANDARD_VERSION=3` in Bitcoin Core v28.0.

**Review by murchandamus — COMMENTED** [review]
Concept ACK and code changes look simple enough. Given that the proposed rules for V3 transactions entail a voluntary topology restriction by the sender on their own transactions, it seems reasonable to me to merge this into the master branch soon

**Review by hernanmarino — COMMENTED** [review]
Concept ACK 7b18f0ed94557a93c69b4f19d85c4fd6fd480464

**Comment by glozow** [comment]
Added constant for https://github.com/bitcoin/bitcoin/pull/29873#discussion_r1610126595
Also added a test that wallet does not signal v3 as it uses the default `CURRENT_VERSION` (this is correct as it wouldn't be safe to make transactions v3 before any of the network has adopted this change). I believe the `nVersion` can be changed via `CCoinControl::m_version`, though I'm not sure if that is accessible so I didn't write a test for it.

Rebased on top of #29873 + master (conflict with #29086)

File: src/policy/v3_policy.h [hunk 1]
```diff
@@ -15,8 +15,9 @@
 #include <set>
 #include <string>
 
-// This module enforces rules for transactions with nVersion=3 ("v3 transactions") which help make
+// This module enforces rules for transactions BIP 431 TRUC transactions (with nVersion=3) which help make
```
Comment (murchandamus) [inline src/policy/v3_policy.h:18]: There is either too many or too few words here:
```suggestion
// This module enforces rules for BIP 431 TRUC transactions (with nVersion=3) which help make
```

File: test/functional/wallet_create_tx.py [hunk 2]
```diff
@@ -106,6 +110,23 @@ def test_create_too_long_mempool_chain(self):
...
+
+        tx_data = wallet_rpc.send(outputs=[{wallet_v3.getnewaddress(): 25}], options={"change_position": 0})
+        wallet_tx_data = wallet_rpc.gettransaction(tx_data["txid"])
+        tx_current_version = tx_from_hex(wallet_tx_data["hex"])
+
+        # While v3 transactions are standard, the CURRENT_VERSION is 2.
+        # This test can be removed if CURRENT_VERSION is changed, and replaced with tests that the
+        # wallet handles v3 rules properly.
+        assert_equal(tx_current_version.nVersion, 2)
+        wallet_v3.unloadwallet()
```
Comment (murchandamus) [inline test/functional/wallet_create_tx.py:128]: Good idea to add this!

**Review by murchandamus — COMMENTED** [review]
Still looks good to me

crACK e41dae322d435cd8b32daf73883b466f30349584

File: src/policy/v3_policy.h (hunk 1 above)
Comment (glozow) [reply src/policy/v3_policy.h]: fixed, thanks

**Comment by glozow** [comment]
Rebased for #30072

File: src/validation.cpp [hunk 3]
```diff
@@ -1017,7 +1017,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
             .descendant_count = maybe_rbf_limits.descendant_count + 1,
             .descendant_size_vbytes = maybe_rbf_limits.descendant_size_vbytes + EXTRA_DESCENDANT_TX_SIZE_LIMIT,
         };
-        if (ws.m_vsize > EXTRA_DESCENDANT_TX_SIZE_LIMIT || ws.m_ptx->nVersion == 3) {
```
Comment (instagibbs) [inline src/validation.cpp:1020]: few remaining spots that I can see:
```
src/test/fuzz/package_eval.cpp:176:                tx_mut.nVersion = fuzzed_data_provider.ConsumeBool() ? 3 : CTransaction::CURRENT_VERSION;
src/test/fuzz/tx_pool.cpp:229:            tx_mut.nVersion = fuzzed_data_provider.ConsumeBool() ? 3 : CTransaction::CURRENT_VERSION;
src/test/txvalidation_tests.cpp:289:        mtx_many_sigops.nVersion = 3;
src/test/util/txmempool.cpp:121:        if (tx_info.tx->nVersion == 3) {
src/test/util/txmempool.cpp:136:                Assert(parents.begin()->get().GetSharedTx()->nVersion == 3);
src/test/util/txmempool.cpp:141:                Assert(parent.get().GetSharedTx()->nVersion != 3);
```

can also use this constant

Comment (glozow) [reply src/validation.cpp:1020]: thanks! fixed these now

**Review by instagibbs — APPROVED** [review]
utACK 30a01134cdec37e7467fcd6eee8b0ae3890a131c

**Review by ismaelsadeeq — COMMENTED** [review]
Concept ACK

File: src/policy/v3_policy.cpp [hunk 4]
```diff
@@ -146,14 +146,14 @@ std::optional<std::string> PackageV3Checks(const CTransactionRef& ptx, int64_t v
     } else {
         // Non-v3 transactions cannot have v3 parents.
```
[the remark this answers is not in the document]
Comment (glozow) [reply src/policy/v3_policy.cpp:147]: Sure, can be in another PR

File: src/policy/v3_policy.h [hunk 5]
```diff
@@ -15,8 +15,9 @@
 #include <set>
 #include <string>
 
-// This module enforces rules for transactions with nVersion=3 ("v3 transactions") which help make
+// This module enforces rules for BIP 431 TRUC transactions (with nVersion=3) which help make
```
[the remark this answers is not in the document]
Comment (murchandamus) [reply src/policy/v3_policy.h:18]: The file name could also be renamed to refer to TRUC transactions instead of v3_policy.

**Review by murchandamus — COMMENTED** [review]
ACK 30a01134cdec37e7467fcd6eee8b0ae3890a131c

**Review by ismaelsadeeq — APPROVED** [review]
ACK 30a01134cdec37e7467fcd6eee8b0ae3890a131c 🛰️

**Comment by sdaftuar** [comment]
utACK 30a01134c

**Comment by achow101** [comment]
ACK 30a01134cdec37e7467fcd6eee8b0ae3890a131c

File: src/policy/v3_policy.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (glozow) [reply src/policy/v3_policy.h:18]: done in #30272

File: src/policy/v3_policy.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (glozow) [reply src/policy/v3_policy.cpp:147]: done in #30272

**Review by 5twelve — APPROVED** [review]

**Pull request merged**